### ISSUE TITLE
rodsadmin user moves folder

### ISSUE THREAD

**Issue by tedgin** [irods/irods#6239]
### iRODS Version, OS and Version
iRODS 4.2.8, CentOS 7

### What did you try to do?
A  user has a collection named `/iplant/home/user/ Ss  /XXXX_vX.X.X_analysis1-2021-10-27-12-19-57.5`. The parent collection ` Ss  ` has a single leading blank and two trailing blanks. The collection contains a child collection named `index`, which contains a data object named `ht2`. Using the Jargon library, the admin user `de-irods` renames the collection to `/iplant/trash/home/de-irods/user/XXXX_vX.X.X_analysis1-2021-10-27-12-19-57.5.0aUcVvx`. It does NOT use the built in move to trash functionality.  It uses a standard move/rename.  Notices that the appends an extension to the collection's original name.

### Expected behavior

I expect the collection and all of its contents to be renamed appropriately, and I expect the physical path of the replica to change similarly.

### Observed behavior (including steps to reproduce, if applicable)

The collection is renamed correctly. However the child collection and data object replica path aren't.  The replica path remains unchaged, but the path to the child collection, `index`,  gets mangled, orphaning it. Here's what's in the ICAT.

```
parent_coll_name | /iplant/home/user/ Ss  //XXXX_vX.X.X_analysis1-2021-10-27-12-19-57.5.0aUcVvx
coll_name        | /iplant/home/user/ Ss  //XXXX_vX.X.X_analysis1-2021-10-27-12-19-57.5.0aUcVvxindex
```

There are no errors in the logs.

**Comment by trel** [comment]
So, is this a Jargon issue?  Or do you think this is in the server?

**Comment by tedgin** [comment]
There might be a Jargon issue too, but the server shouldn't allow children to become orphans in the ICAT.

**Comment by trel** [comment]
Agreed - will try to reproduce.   Is the rodsadmin part necessary?   Can a rodsuser move something into trash manually and see the same problem?

**Comment by tedgin** [comment]
I'm not certain.  The DE doesn't provide a way for the user to move something to trash as the user. 

This issue has existed for a while, but this is the first time I observed it in 4.2.8. When I observed it in iRODS 4.1.10, I didn't report it, because iRODS 4.1.10 was EOL at the time. 

The destination collection doesn't have to be trash. I've seen the same corruption occur with other DE moves. Nowadays the DE performs these moves by proxying as the user. I'm not certain if that was the case when I last saw the issue in iRODS 4.1.10. We'll see if we can reproduce this.

**Comment by trel** [comment]
Ah, so trash itself may not even be relevant.    Okay, probably the spaces are not being escaped/handled somewhere and getting parsed incorrectly.

**Comment by tsmeele** [comment]
We have found similar mangled collection name cases, we run Centos7 and iRODS 4.2.12.  We do not use Jargon, and parent collection does not start/end with spaces.  So far, attempts to reproduce using icommands or msi have failed.  

Example:
```
iquest "select COLL_NAME,COLL_PARENT_NAME where COLL_NAME = '/nluu5p/home/REDACTED Put in Cupboard April 1930akelos2'"
COLL_NAME = /nluu5p/home/REDACTED Put in Cupboard April 1930akelos2
COLL_PARENT_NAME = /nluu5p/home/REDACTED Put in Cupboard April 1930
```

**Comment by trel** [comment]
got it - thanks.

```
irods$ iadmin asq "select coll_name from r_coll_main where not starts_with(coll_name, parent_coll_name||'/')" issue6239
irods$ iquest --sql issue6239
/
----
/tempZone
```

**Comment by tsmeele** [comment]
The above results are false positives, always returned by the query as their parent colls are special.   So your system does not yet exhibit the issue.   The cases that i have found so far are fairly recent, which suggests that the issue might have been introduced in 4.2.12.

**Comment by trel** [comment]
Right - understood.

Here I'm mangling a coll_name directly... and then seeing the mangled entry via your query.

```
irods$ imkdir -p one/two
irods$ psql ICAT -c "update r_coll_main set coll_name = '/tempZone/home/rods/onetwo' where coll_id = (select coll_id from r_coll_main where coll_name = '/tempZone/home/rods/one/two')"
UPDATE 1
irods$ iquest --sql issue6239
/
----
/tempZone
----
/tempZone/home/rods/onetwo
```

**Comment by tsmeele** [comment]
Finally I have been able to reproduce the issue, steps are:
```
$ imkdir -p 'béta alpha/gamma'
$ ils
/tempZone/home/ton/test:
  C- /tempZone/home/ton/test/béta alpha
$ ils 'béta alpha'
/tempZone/home/ton/test/béta alpha:
  C- /tempZone/home/ton/test/béta alpha/gamma
$ imv 'béta alpha' 'béta zeta'
$ ils
/tempZone/home/ton/test:
  C- /tempZone/home/ton/test/béta zeta
$ ils 'béta zeta'
/tempZone/home/ton/test/béta zeta:
  C- /tempZone/home/ton/test/béta zetagamma
```
Note that the collection has a name that includes a multi-byte unicode character and a space. Renaming this collection mangles its subcollection.

**Comment by tsmeele** [comment]
This might be related to the issue https://github.com/irods/irods/blob/8b515bf4c4c06dd6c9c7484584b35ddbaa37a0f5/plugins/database/src/db_plugin.cpp#L11375
strlen returns #bytes, whereas the postgresql substr function expects #characters

**Comment by SwooshyCueb** [comment]
I'm not sure how postgres counts characters, but the example code on this page implements multibyte character aware character counting using `mblen`: https://en.cppreference.com/w/cpp/string/multibyte/mblen

Linking to #8055

**Comment by tsmeele** [comment]
i have created a pull request for having RDBMS do the length calculation (again), to avoid misinterpretation.

**Comment by trel** [comment]
PR here... https://github.com/irods/irods/pull/8076

**Comment by alanking** [comment]
Just making a note here regarding trying to reproduce the issue in OP...

In 4.3.3, I do not see orphaned contents, but the name of the collection which appears in the trash is similar to what is shown in OP minus the extra bits added to the end (that is, the `  Ss   ` member is omitted from the path, but there's no "mangling" at the end). Note: Even GitHub's parser for code snippets is chopping off a space on each end of ` Ss  ` so I had to add extras in the parenthetical statement...

**Comment by tsmeele** [comment]
The mangling is not related to spaces in the name (that was the initial assumption).  Try reproduce with an accented character, for instance "ê"

Met vriendelijke groet,

Ton
________________________________

**Comment by alanking** [comment]
> The mangling is not related to spaces in the name (that was the initial assumption).  Try reproduce with an accented character, for instance "ê"

Yes, understood. Was just making a note about the scenario in OP. I'll try to reproduce and add a test for the multi-byte characters as well.

Actually... the missing component in the path is the expected behavior for what I did.

I renamed this collection like this:
```
imv /tempZone/home/alan/\ Ss\ \ /XXXX_vX.X.X_analysis1-2021-10-27-12-19-57.5 /tempZone/trash/home/rods/
```
The actual collection being renamed would not carry over the parent collection (`/tempZone/home/alan/ Ss  `) to the destination path. So I think that part is all accounted for. I'll add a test for that anyway.

**Comment by alanking** [comment]
@tsmeele - I was able to reproduce the multi-byte character issue as described here: https://github.com/irods/irods/issues/6239#issuecomment-2535223519 However... the multi-byte character issue remains even with the changes made in https://github.com/irods/irods/pull/8076.

@tedgin - The issue in OP is not reproducible in 4.3.3, so this issue will now be focused around the multi-byte character problem.

**Comment by alanking** [comment]
Good work, @tsmeele. Closing this one.
